File: src/main/java/org/ymxr/Layout.java
package org.ymxr;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.TreeSet;
import org.ymxs.tool.Tool;

/**
 * The record of BINARIES.md 6: a file this repository writes read back
 * and reported part by part, one line of JSON a part. The reader here
 * reads the file alone, as an implementer of the document does, so a
 * record that differs from the reference is the document read two ways.
 *
 * <p>{@code ymxr-layout} (tools.md 9.5) writes it.
 */
final class Layout {

    private Layout() {
    }

    /** The tool: a file on standard input, its record on standard
     *  output. */
    public static void main(String[] args) {
        List<String> flags = new ArrayList<>(Arrays.asList(args));
        Tool tool = Tool.of("ymxr-layout", flags);
        byte[] file = tool.bytes();
        Report report = new Report(tool.reports());
        String record;
        try {
            record = of(file);
        } catch (IllegalArgumentException | ArrayIndexOutOfBoundsException wrong) {
            throw tool.wrong(Tool.WRONG, wrong instanceof IllegalArgumentException
                    ? String.valueOf(wrong.getMessage())
                    : "the record runs past the file's " + file.length + " bytes");
        }
        List<String> lines = record.lines().toList();
        report.say("the file: " + file.length + " bytes, kind " + kind(file));
        for (String part : List.of("tag", "tune", "source", "image")) {
            long of = lines.stream().filter(line -> line.contains("\"part\":\"" + part + "\""))
                    .count();
            if (of > 0) {
                report.row("the " + part + (of == 1 ? "" : "s"), String.valueOf(of));
            }
        }
        tool.report(lines.size() + (lines.size() == 1 ? " line" : " lines"));
        Out.write(tool, record.getBytes(StandardCharsets.US_ASCII));
    }

    /** The record of the file: 6.1's first line, then the lines of 6.2
     *  to 6.5 by kind, each ended by a line feed. */
    static String of(byte[] file) {
        StringBuilder out = new StringBuilder();
        String kind = kind(file);
        out.append("{\"kind\":\"").append(kind).append("\",\"bytes\":").append(file.length)
                .append("}\n");
        switch (kind) {
            case "multi" -> multi(out, file);
            case "bound" -> bound(out, file);
            case "program" -> program(out, file);
            default -> sndh(out, file, 0, file.length);
        }
        return out.toString();
    }

    /** The kind of 6.1, read off the file's first bytes. */
    private static String kind(byte[] file) {
        if (ascii(file, 0, 4).equals("YMXM")) {
            return "multi";
        }
        if (ascii(file, 0, 4).equals("YMXB")) {
            return "bound";
        }
        if (file.length >= 2 && word(file, 0) == Prg.PRG_MAGIC) {
            return "program";
        }
        if (ascii(file, 12, 4).equals("SNDH")) {
            return "sndh";
        }
        throw new IllegalArgumentException("not a file BINARIES.md defines: no YMXM, YMXB,"
                + " $601A or SNDH");
    }

    /** 6.2: the header, then a line a tune. */
    private static void multi(StringBuilder out, byte[] file) {
        int tunes = word(file, 6);
        out.append("{\"part\":\"header\",\"version\":").append(word(file, 4))
                .append(",\"tunes\":").append(tunes).append("}\n");
        int name = 8 + 8 * tunes;
        for (int i = 0; i < tunes; i++) {
            int to = zero(file, name);
            out.append("{\"part\":\"tune\",\"number\":").append(i + 1)
                    .append(",\"at\":").append(Tune.getLong(file, 8 + 8 * i))
                    .append(",\"bytes\":").append(Tune.getLong(file, 12 + 8 * i))
                    .append(",\"name\":").append(text(new String(file, name, to - name,
                            StandardCharsets.UTF_8)))
                    .append("}\n");
            name = to + 1;
        }
    }

    /** 6.3: the header, a line a source, and the image of a bound tune
     *  written alone. */
    private static void bound(StringBuilder out, byte[] file) {
        out.append("{\"part\":\"header\"").append(fields(file, 0)).append("}\n");
        sources(out, file, 0);
        int image = Tune.getLong(file, 16);
        if (image > 0) {
            out.append("{\"part\":\"image\",\"at\":").append(image)
                    .append(",\"bytes\":").append(file.length - image).append("}\n");
        }
    }

    /** 6.5: the PRG header, the stub, where the SNDH file begins, and
     *  that file's lines. The SNDH file ends at 28 plus the long at 2,
     *  the relocation table standing after it. */
    private static void program(StringBuilder out, byte[] file) {
        int at = 28;
        while (!ascii(file, at + 12, 4).equals("SNDH")) {
            at += 2;
        }
        out.append("{\"part\":\"prg\",\"text\":").append(Tune.getLong(file, 2)).append("}\n");
        int version = word(file, 28 + 8);
        out.append("{\"part\":\"stub\",\"at\":28,\"bytes\":").append(at - 28)
                .append(",\"version\":").append(version)
                .append(",\"subtunes\":").append(word(file, 28 + 10))
                .append(",\"flags\":").append(word(file, 28 + 12))
                .append(",\"rate\":").append(word(file, 28 + 14))
                .append(",\"rows\":").append(Tune.getLong(file, 28 + 16))
                .append(",\"core\":").append(at + Tune.getLong(file, 28 + 20));
        if (version >= 2) {
            // the timer the stub arms (4.10), which a program of an
            // earlier version has no field for
            out.append(",\"prescaler\":").append(word(file, 28 + Prg.STUB_PRESCALER_AT))
                    .append(",\"count\":").append(word(file, 28 + Prg.STUB_COUNT_AT))
                    .append(",\"ticks\":").append(word(file, 28 + Prg.STUB_TICKS_AT));
        }
        out.append("}\n");
        out.append("{\"part\":\"sndh\",\"at\":").append(at).append("}\n");
        sndh(out, file, at, 28 + Tune.getLong(file, 2));
    }

    /** 6.4: the entry triple, the tags, the core, the subtune table, a
     *  subtune and its sources a line each, a line an image, and the
     *  workspace. The SNDH file runs from {@code from} to {@code ends},
     *  and every offset counts from the file's first byte. */
    private static void sndh(StringBuilder out, byte[] file, int from, int ends) {
        out.append("{\"part\":\"entry\",\"to\":[");
        for (int i = 0; i < 3; i++) {
            out.append(i == 0 ? "" : ",").append(from + 4 * i + 2
                    + (short) word(file, from + 4 * i + 2));
        }
        out.append("]}\n");
        int hdns = tags(out, file, from);
        int core = Sndh.even(hdns + 4);
        int table = core + Tune.getLong(file, core + 28);
        int work = core + Tune.getLong(file, core + 32);
        out.append("{\"part\":\"core\",\"at\":").append(core)
                .append(",\"version\":").append(word(file, core + 16))
                .append(",\"binds\":").append(word(file, core + 18))
                .append(",\"fixed\":").append(word(file, core + 20))
                .append(",\"flags\":").append(word(file, core + 22))
                .append(",\"state\":").append(core + word(file, core + 24))
                .append(",\"subtunes\":").append(table)
                .append(",\"work\":").append(work)
                .append("}\n");
        int n = word(file, table);
        List<Integer> tunes = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            tunes.add(core + Tune.getLong(file, table + 2 + 4 * i));
        }
        out.append("{\"part\":\"subtunes\",\"at\":").append(table).append(",\"tunes\":[");
        for (int i = 0; i < n; i++) {
            out.append(i == 0 ? "" : ",").append(tunes.get(i));
        }
        out.append("]}\n");
        TreeSet<Integer> images = new TreeSet<>();
        for (int at : tunes) {
            images.add(at + Tune.getLong(file, at + 16));
        }
        for (int i = 0; i < n; i++) {
            int at = tunes.get(i);
            int upto = i + 1 < n ? tunes.get(i + 1) : images.first();
            out.append("{\"part\":\"tune\",\"number\":").append(i + 1)
                    .append(",\"at\":").append(at)
                    .append(",\"bytes\":").append(upto - at)
                    .append(fields(file, at)).append("}\n");
            sources(out, file, at);
        }
        int number = 1;
        for (int image : images) {
            out.append("{\"part\":\"image\",\"number\":").append(number++)
                    .append(",\"at\":").append(image).append("}\n");
        }
        out.append("{\"part\":\"workspace\",\"at\":").append(work)
                .append(",\"bytes\":").append(ends - work).append("}\n");
    }

    /** The tags of 3.2, a line each from 16 to HDNS, and where HDNS
     *  stands. */
    private static int tags(StringBuilder out, byte[] file, int from) {
        int at = from + 16;
        int subtunes = 0;
        while (true) {
            if (file[at] == 0) {
                at++;
                continue;
            }
            String name = ascii(file, at, 4);
            if (name.equals("HDNS")) {
                out.append("{\"part\":\"tag\",\"name\":\"HDNS\",\"at\":").append(at)
                        .append("}\n");
                return at;
            }
            if (name.startsWith("##")) {
                subtunes = Integer.parseInt(name.substring(2));
                out.append("{\"part\":\"tag\",\"name\":\"##\",\"at\":").append(at)
                        .append(",\"subtunes\":").append(subtunes).append("}\n");
                at += 5;
            } else if (name.startsWith(Sndh.TIMER_C_CLOCK) || name.startsWith(Sndh.VBL_CLOCK)) {
                int to = zero(file, at + 2);
                out.append("{\"part\":\"tag\",\"name\":\"").append(name, 0, 2)
                        .append("\",\"at\":").append(at).append(",\"rate\":")
                        .append(Integer.parseInt(ascii(file, at + 2, to - at - 2)))
                        .append("}\n");
                at = to + 1;
            } else if (name.equals("FRMS")) {
                out.append("{\"part\":\"tag\",\"name\":\"FRMS\",\"at\":").append(at)
                        .append(",\"frames\":[");
                for (int i = 0; i < subtunes; i++) {
                    out.append(i == 0 ? "" : ",").append(Tune.getLong(file, at + 4 + 4 * i));
                }
                out.append("]}\n");
                at += 4 + 4 * subtunes;
            } else if (name.equals("!#SN")) {
                out.append("{\"part\":\"tag\",\"name\":\"!#SN\",\"at\":").append(at)
                        .append(",\"names\":[");
                int named = at + 4 + 2 * subtunes;
                for (int i = 0; i < subtunes; i++) {
                    int to = zero(file, named);
                    out.append(i == 0 ? "" : ",").append(text(new String(file, named,
                            to - named, StandardCharsets.UTF_8)));
                    named = to + 1;
                }
                out.append("]}\n");
                at = named;
            } else {
                int to = zero(file, at + 4);
                out.append("{\"part\":\"tag\",\"name\":\"").append(name).append("\",\"at\":")
                        .append(at).append(",\"text\":")
                        .append(text(ascii(file, at + 4, to - at - 4))).append("}\n");
                at = to + 1;
            }
        }
    }

    /** A line a source of the bound tune at {@code at}, in index
     *  order. */
    private static void sources(StringBuilder out, byte[] file, int at) {
        int sources = file[at + 9] & 0xFF;
        for (int i = 0; i < sources; i++) {
            out.append("{\"part\":\"source\",\"number\":").append(i + 1)
                    .append(",\"at\":").append(at + Tune.getLong(file, at + 24 + 4 * i))
                    .append("}\n");
        }
    }

    /** The fields of 1.2 a record reports of the bound tune at
     *  {@code at}, its image resolved against the file. */
    private static String fields(byte[] file, int at) {
        return ",\"version\":" + word(file, at + 4)
                + ",\"rate\":" + word(file, at + 6)
                + ",\"effects\":" + (file[at + 8] & 0xFF)
                + ",\"sources\":" + (file[at + 9] & 0xFF)
                + ",\"state\":" + Tune.getLong(file, at + 12)
                + ",\"image\":" + (at + Tune.getLong(file, at + 16))
                + ",\"table\":" + Tune.getLong(file, at + 20);
    }

    /** Where the next zero byte from {@code from} stands. */
    private static int zero(byte[] file, int from) {
        int at = from;
        while (file[at] != 0) {
            at++;
        }
        return at;
    }

    /** JSON text: the quotes, a backslash before a quote and a
     *  backslash, and every character above $7E escaped, so the record
     *  reads US-ASCII (6.1). A byte sequence outside UTF-8 reads as
     *  U+FFFD, the replacement character, and escapes as $FFFD. */
    private static String text(String said) {
        StringBuilder out = new StringBuilder("\"");
        for (int i = 0; i < said.length(); i++) {
            char c = said.charAt(i);
            if (c == '"' || c == '\\') {
                out.append('\\').append(c);
            } else if (c > 0x7E) {
                out.append(String.format("\\u%04x", (int) c));
            } else {
                out.append(c);
            }
        }
        return out.append('"').toString();
    }

    private static String ascii(byte[] file, int at, int bytes) {
        return at + bytes > file.length ? ""
                : new String(file, at, bytes, StandardCharsets.US_ASCII);
    }

    private static int word(byte[] file, int at) {
        return Tune.getWord(file, at);
    }
}

File: go/sndh/layout.go
package sndh

import (
	"bytes"
	"encoding/json"
	"fmt"
	"runtime"
	"sort"
	"strconv"
	"strings"

	"github.com/odipar/ymxr/go/ymxr"
)

// Layout is the record of BINARIES.md 6: a file this repository writes
// read back and reported part by part, one line of JSON a part. The
// reader here reads the file alone, as an implementer of the document
// does, so a record that differs from the reference is the document read
// two ways. A read past the file's end is the error of tools.md 9.7,
// where the Java tree catches the same read.
func Layout(file []byte) (record string, err error) {
	defer func() {
		if r := recover(); r != nil {
			wrong, ok := r.(runtime.Error)
			if !ok || !(strings.Contains(wrong.Error(), "index out of range") ||
				strings.Contains(wrong.Error(), "slice bounds out of range")) {
				panic(r)
			}
			record, err = "", fmt.Errorf("the record runs past the file's %d bytes",
				len(file))
		}
	}()
	kind, err := kindOf(file)
	if err != nil {
		return "", err
	}
	var out bytes.Buffer
	fmt.Fprintf(&out, "{\"kind\":\"%s\",\"bytes\":%d}\n", kind, len(file))
	switch kind {
	case "multi":
		layoutMulti(&out, file)
	case "bound":
		layoutBound(&out, file)
	case "program":
		layoutProgram(&out, file)
	default:
		layoutSndh(&out, file, 0, len(file))
	}
	return out.String(), nil
}

// kindOf is the kind of 6.1, read off the file's first bytes, the four in
// the order the clause lists them.
func kindOf(file []byte) (string, error) {
	if ascii(file, 0, 4) == "YMXM" {
		return "multi", nil
	}
	if ascii(file, 0, 4) == "YMXB" {
		return "bound", nil
	}
	if len(file) >= 2 && ymxr.GetWord(file, 0) == PrgMagic {
		return "program", nil
	}
	if ascii(file, 12, 4) == "SNDH" {
		return "sndh", nil
	}
	return "", fmt.Errorf("not a file BINARIES.md defines: no YMXM, YMXB, $601A or SNDH")
}

// layoutMulti is 6.2: the header, then a line a tune.
func layoutMulti(out *bytes.Buffer, file []byte) {
	tunes := ymxr.GetWord(file, 6)
	fmt.Fprintf(out, "{\"part\":\"header\",\"version\":%d,\"tunes\":%d}\n",
		ymxr.GetWord(file, 4), tunes)
	name := 8 + 8*tunes
	for i := 0; i < tunes; i++ {
		to := zeroFrom(file, name)
		fmt.Fprintf(out, "{\"part\":\"tune\",\"number\":%d,\"at\":%d,\"bytes\":%d,"+
			"\"name\":%s}\n", i+1, ymxr.GetLong(file, 8+8*i),
			ymxr.GetLong(file, 12+8*i), jsonText(string(file[name:to])))
		name = to + 1
	}
}

// layoutBound is 6.3: the header, a line a source, and the image of a
// bound tune written alone.
func layoutBound(out *bytes.Buffer, file []byte) {
	fmt.Fprintf(out, "{\"part\":\"header\"%s}\n", boundFields(file, 0))
	layoutSources(out, file, 0)
	if image := ymxr.GetLong(file, 16); image > 0 {
		fmt.Fprintf(out, "{\"part\":\"image\",\"at\":%d,\"bytes\":%d}\n",
			image, len(file)-image)
	}
}

// layoutProgram is 6.5: the PRG header, the stub, where the SNDH file
// begins, and that file's lines. The SNDH file ends at 28 plus the long
// at 2, the relocation table after it.
func layoutProgram(out *bytes.Buffer, file []byte) {
	at := 28
	// the first SNDH on an even offset; a file cut before it ends the
	// search at its end, and the tags below read past it (tools.md 9.7)
	for at+16 <= len(file) && ascii(file, at+12, 4) != "SNDH" {
		at += 2
	}
	fmt.Fprintf(out, "{\"part\":\"prg\",\"text\":%d}\n", ymxr.GetLong(file, 2))
	version := ymxr.GetWord(file, 28+8)
	fmt.Fprintf(out, "{\"part\":\"stub\",\"at\":28,\"bytes\":%d,\"version\":%d,"+
		"\"subtunes\":%d,\"flags\":%d,\"rate\":%d,\"rows\":%d,\"core\":%d",
		at-28, version, ymxr.GetWord(file, 28+10),
		ymxr.GetWord(file, 28+12), ymxr.GetWord(file, 28+14),
		ymxr.GetLong(file, 28+16), at+ymxr.GetLong(file, 28+20))
	if version >= 2 {
		// the timer the stub arms (4.10), which a program of an earlier
		// version has no field for
		fmt.Fprintf(out, ",\"prescaler\":%d,\"count\":%d,\"ticks\":%d",
			ymxr.GetWord(file, 28+stubPrescalerAt), ymxr.GetWord(file, 28+stubCountAt),
			ymxr.GetWord(file, 28+stubTicksAt))
	}
	out.WriteString("}\n")
	fmt.Fprintf(out, "{\"part\":\"sndh\",\"at\":%d}\n", at)
	layoutSndh(out, file, at, 28+ymxr.GetLong(file, 2))
}

// layoutSndh is 6.4: the entry triple, the tags, the core, the subtune
// table, a subtune and its sources a line each, a line an image, and the
// workspace. The SNDH file runs from `from` to `ends`, and every offset
// counts from the file's first byte.
func layoutSndh(out *bytes.Buffer, file []byte, from, ends int) {
	out.WriteString("{\"part\":\"entry\",\"to\":[")
	for i := 0; i < 3; i++ {
		if i > 0 {
			out.WriteString(",")
		}
		fmt.Fprintf(out, "%d", from+4*i+2+int(int16(ymxr.GetWord(file, from+4*i+2))))
	}
	out.WriteString("]}\n")
	hdns := layoutTags(out, file, from)
	core := even(hdns + 4)
	table := core + ymxr.GetLong(file, core+28)
	work := core + ymxr.GetLong(file, core+32)
	fmt.Fprintf(out, "{\"part\":\"core\",\"at\":%d,\"version\":%d,\"binds\":%d,"+
		"\"fixed\":%d,\"flags\":%d,\"state\":%d,\"subtunetable\":%d,\"work\":%d}\n",
		core, ymxr.GetWord(file, core+16), ymxr.GetWord(file, core+18),
		ymxr.GetWord(file, core+20), ymxr.GetWord(file, core+22),
		core+ymxr.GetWord(file, core+24), table, work)
	n := ymxr.GetWord(file, table)
	tunes := make([]int, n)
	for i := 0; i < n; i++ {
		tunes[i] = core + ymxr.GetLong(file, table+2+4*i)
	}
	fmt.Fprintf(out, "{\"part\":\"subtunes\",\"at\":%d,\"tunes\":[", table)
	for i, at := range tunes {
		if i > 0 {
			out.WriteString(",")
		}
		fmt.Fprintf(out, "%d", at)
	}
	out.WriteString("]}\n")
	var images []int
	for _, at := range tunes {
		image := at + ymxr.GetLong(file, at+16)
		if !has(images, image) {
			images = append(images, image)
		}
	}
	sort.Ints(images)
	for i, at := range tunes {
		upto := images[0]
		if i+1 < n {
			upto = tunes[i+1]
		}
		fmt.Fprintf(out, "{\"part\":\"tune\",\"number\":%d,\"at\":%d,\"bytes\":%d%s}\n",
			i+1, at, upto-at, boundFields(file, at))
		layoutSources(out, file, at)
	}
	for i, at := range images {
		fmt.Fprintf(out, "{\"part\":\"image\",\"number\":%d,\"at\":%d}\n", i+1, at)
	}
	fmt.Fprintf(out, "{\"part\":\"workspace\",\"at\":%d,\"bytes\":%d}\n", work, ends-work)
}

// layoutTags is the tags of 3.2, a line each from 16 to HDNS, and where
// HDNS is.
func layoutTags(out *bytes.Buffer, file []byte, from int) int {
	at := from + 16
	subtunes := 0
	for {
		if file[at] == 0 {
			at++
			continue
		}
		name := ascii(file, at, 4)
		switch {
		case name == "HDNS":
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"HDNS\",\"at\":%d}\n", at)
			return at
		case name[:2] == "##":
			subtunes, _ = strconv.Atoi(name[2:])
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"##\",\"at\":%d,"+
				"\"count\":%d}\n", at, subtunes)
			at += 5
		case name[:2] == ClockTimerC || name[:2] == ClockVBL:
			to := zeroFrom(file, at+2)
			rate := 0
			for i := at + 2; i < to && file[i] >= '0' && file[i] <= '9'; i++ {
				rate = rate*10 + int(file[i]-'0')
			}
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"%s\",\"at\":%d,"+
				"\"rate\":%d}\n", name[:2], at, rate)
			at = to + 1
		case name == "FRMS":
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"FRMS\",\"at\":%d,"+
				"\"frames\":[", at)
			for i := 0; i < subtunes; i++ {
				if i > 0 {
					out.WriteString(",")
				}
				fmt.Fprintf(out, "%d", ymxr.GetLong(file, at+4+4*i))
			}
			out.WriteString("]}\n")
			at += 4 + 4*subtunes
		case name == "TIME":
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"TIME\",\"at\":%d,"+
				"\"seconds\":[", at)
			for i := 0; i < subtunes; i++ {
				if i > 0 {
					out.WriteString(",")
				}
				fmt.Fprintf(out, "%d", ymxr.GetWord(file, at+4+2*i))
			}
			out.WriteString("]}\n")
			at += 4 + 2*subtunes
		case name == "!#SN":
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"!#SN\",\"at\":%d,"+
				"\"names\":[", at)
			named := at + 4 + 2*subtunes
			for i := 0; i < subtunes; i++ {
				if i > 0 {
					out.WriteString(",")
				}
				to := zeroFrom(file, named)
				out.WriteString(jsonText(string(file[named:to])))
				named = to + 1
			}
			out.WriteString("]}\n")
			at = named
		default:
			to := zeroFrom(file, at+4)
			fmt.Fprintf(out, "{\"part\":\"tag\",\"name\":\"%s\",\"at\":%d,"+
				"\"text\":%s}\n", name, at, jsonText(string(file[at+4:to])))
			at = to + 1
		}
	}
}

// layoutSources is a line a source of the bound tune at `at`, in index
// order.
func layoutSources(out *bytes.Buffer, file []byte, at int) {
	for i := 0; i < int(file[at+9]); i++ {
		fmt.Fprintf(out, "{\"part\":\"source\",\"number\":%d,\"at\":%d}\n",
			i+1, at+ymxr.GetLong(file, at+24+4*i))
	}
}

// boundFields is the fields of 1.2 a record reports of the bound tune at
// `at`, its image resolved against the file.
func boundFields(file []byte, at int) string {
	return fmt.Sprintf(",\"version\":%d,\"rate\":%d,\"effects\":%d,\"sources\":%d,"+
		"\"stateblock\":%d,\"image\":%d,\"table\":%d",
		ymxr.GetWord(file, at+4), ymxr.GetWord(file, at+6), file[at+8], file[at+9],
		ymxr.GetLong(file, at+12), at+ymxr.GetLong(file, at+16),
		ymxr.GetLong(file, at+20))
}

// zeroFrom is where the next zero byte from `from` is.
func zeroFrom(file []byte, from int) int {
	at := from
	for file[at] != 0 {
		at++
	}
	return at
}

// jsonText is JSON text: the quotes, the escapes JSON defines, and a
// character above $7E escaped as \uXXXX.
func jsonText(said string) string {
	out, err := json.Marshal(said)
	if err != nil {
		return "\"\""
	}
	var escaped bytes.Buffer
	for _, r := range string(out) {
		if r > 0x7E {
			fmt.Fprintf(&escaped, "\\u%04x", r)
		} else {
			escaped.WriteRune(r)
		}
	}
	return escaped.String()
}

func ascii(file []byte, at, bytes int) string {
	if at < 0 || at+bytes > len(file) {
		return ""
	}
	return string(file[at : at+bytes])
}

func has(of []int, one int) bool {
	for _, each := range of {
		if each == one {
			return true
		}
	}
	return false
}
